File: AlgorithmLens_Cowork/mobile/modules/broadcast/ios/BroadcastModule.swift
import ExpoModulesCore
import ReplayKit

/**
 * BroadcastModule — Expo native module bridge for iOS ReplayKit broadcast.
 *
 * This module provides the React Native interface to query and manage
 * ReplayKit screen broadcasts. The actual frame capture happens in the
 * BroadcastExtension (a separate iOS target that runs in its own process).
 *
 * Communication between the main app and the extension happens through
 * a shared App Group container (group.com.algorithmlens.broadcast).
 *
 * Responsibilities:
 * - Query broadcast availability and status
 * - Read session metadata written by the extension
 * - Read frame metadata and frame file paths
 * - Collect captured frames for upload to analysis pipeline
 * - Clean up shared container after processing
 * - Emit status change events to React Native
 *
 * The RPSystemBroadcastPickerView (system broadcast trigger) is handled
 * at the React Native layer via a native UIView component.
 */
public class BroadcastModule: Module {

    private let appGroupID = "group.com.algorithmlens.broadcast"
    private var statusPollTimer: Timer?

    public func definition() -> ModuleDefinition {
        Name("ExpoBroadcast")

        // Emit events to JS when broadcast status changes
        Events("onStatusChange", "onFrameCountUpdate")

        // MARK: - Availability & Status

        /// Returns whether ReplayKit broadcast is available on this device.
        /// Requires iOS 12+ and the broadcast extension to be configured.
        Function("isAvailable") { () -> Bool in
            if #available(iOS 12.0, *) {
                return true
            }
            return false
        }

        /// Returns the current broadcast session status from the shared container.
        /// Reads the session_metadata.json written by the broadcast extension.
        Function("getStatus") { () -> [String: Any] in
            guard let containerURL = FileManager.default.containerURL(
                forSecurityApplicationGroupIdentifier: self.appGroupID
            ) else {
                return ["status": "IDLE", "error": "Shared container not available"]
            }

            let metadataURL = containerURL.appendingPathComponent("session_metadata.json")
            guard FileManager.default.fileExists(atPath: metadataURL.path),
                  let data = try? Data(contentsOf: metadataURL),
                  let metadata = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
                return ["status": "IDLE"]
            }

            return metadata
        }

        /// Returns the shared container path for debugging/diagnostics.
        Function("getSharedContainerPath") { () -> String? in
            let containerURL = FileManager.default.containerURL(
                forSecurityApplicationGroupIdentifier: self.appGroupID
            )
            return containerURL?.path
        }

        // MARK: - Frame Data

        /// Returns the number of unique JPEG frames stored in the shared container.
        Function("getFrameCount") { () -> Int in
            guard let containerURL = FileManager.default.containerURL(
                forSecurityApplicationGroupIdentifier: self.appGroupID
            ) else {
                return 0
            }
            let framesDir = containerURL.appendingPathComponent("frames")
            let contents = try? FileManager.default.contentsOfDirectory(atPath: framesDir.path)
            return contents?.filter { $0.hasSuffix(".jpg") }.count ?? 0
        }

        /// Returns the frame metadata array written by the broadcast extension.
        /// Each entry contains: frame_id, filename, captured_at, size_bytes,
        /// width, height, ocr_text, ocr_confidence, is_unique, frame_number.
        Function("getFrameMetadata") { () -> [[String: Any]] in
            guard let containerURL = FileManager.default.containerURL(
                forSecurityApplicationGroupIdentifier: self.appGroupID
            ) else {
                return []
            }

            let metadataURL = containerURL.appendingPathComponent("frame_metadata.json")
            guard FileManager.default.fileExists(atPath: metadataURL.path),
                  let data = try? Data(contentsOf: metadataURL),
                  let frames = try? JSONSerialization.jsonObject(with: data) as? [[String: Any]] else {
                return []
            }

            return frames
        }

        /// Returns the full file paths for all captured JPEG frames,
        /// sorted by filename (which includes the capture timestamp).
        Function("getFramePaths") { () -> [String] in
            guard let containerURL = FileManager.default.containerURL(
                forSecurityApplicationGroupIdentifier: self.appGroupID
            ) else {
                return []
            }

            let framesDir = containerURL.appendingPathComponent("frames")
            guard let contents = try? FileManager.default.contentsOfDirectory(atPath: framesDir.path) else {
                return []
            }

            return contents
                .filter { $0.hasSuffix(".jpg") }
                .sorted()
                .map { framesDir.appendingPathComponent($0).path }
        }

        /// Reads a single frame's JPEG data as a base64-encoded string.
        /// Used for sending frames to the Gemini analysis pipeline.
        Function("getFrameBase64") { (filename: String) -> String? in
            guard let containerURL = FileManager.default.containerURL(
                forSecurityApplicationGroupIdentifier: self.appGroupID
            ) else {
                return nil
            }

            let fileURL = containerURL
                .appendingPathComponent("frames")
                .appendingPathComponent(filename)

            guard let data = try? Data(contentsOf: fileURL) else {
                return nil
            }

            return data.base64EncodedString()
        }

        // MARK: - Session Management

        /// Starts polling the shared container for status changes.
        /// Call this when the user initiates a broadcast session.
        /// The module reads session_metadata.json every second and emits
        /// onStatusChange events when the status changes.
        Function("startStatusPolling") { () in
            self.stopPolling()

            var lastStatus = ""
            var lastFrameCount = 0

            self.statusPollTimer = Timer.scheduledTimer(withTimeInterval: 1.0, repeats: true) { [weak self] _ in
                guard let self = self else { return }

                guard let containerURL = FileManager.default.containerURL(
                    forSecurityApplicationGroupIdentifier: self.appGroupID
                ) else { return }

                let metadataURL = containerURL.appendingPathComponent("session_metadata.json")

                guard FileManager.default.fileExists(atPath: metadataURL.path),
                      let data = try? Data(contentsOf: metadataURL),
                      let metadata = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
                    return
                }

                let currentStatus = metadata["status"] as? String ?? "IDLE"
                let currentFrames = metadata["frames_unique"] as? Int ?? 0

                // Emit status change event
                if currentStatus != lastStatus {
                    lastStatus = currentStatus
                    self.sendEvent("onStatusChange", [
                        "status": currentStatus,
                        "metadata": metadata
                    ])

                    // Auto-stop polling when session completes or fails
                    if currentStatus == "COMPLETE" || currentStatus == "FAILED" {
                        self.stopPolling()
                    }
                }

                // Emit frame count update
                if currentFrames != lastFrameCount {
                    lastFrameCount = currentFrames
                    self.sendEvent("onFrameCountUpdate", [
                        "frameCount": currentFrames,
                        "framesTotal": metadata["frames_captured"] as? Int ?? 0
                    ])
                }
            }
        }

        /// Stops polling for status changes.
        Function("stopStatusPolling") { () in
            self.stopPolling()
        }

        /// Prepares the shared container for a new broadcast session.
        /// Cleans up any stale data from previous sessions.
        AsyncFunction("prepareSession") { (promise: Promise) in
            guard let containerURL = FileManager.default.containerURL(
                forSecurityApplicationGroupIdentifier: self.appGroupID
            ) else {
                promise.reject("CONTAINER_ERROR", "Shared container not available")
                return
            }

            do {
                // Remove stale frames
                let framesDir = containerURL.appendingPathComponent("frames")
                if FileManager.default.fileExists(atPath: framesDir.path) {
                    try FileManager.default.removeItem(at: framesDir)
                }
                try FileManager.default.createDirectory(
                    at: framesDir,
                    withIntermediateDirectories: true
                )

                // Remove stale metadata
                let sessionMetaURL = containerURL.appendingPathComponent("session_metadata.json")
                if FileManager.default.fileExists(atPath: sessionMetaURL.path) {
                    try FileManager.default.removeItem(at: sessionMetaURL)
                }

                let frameMetaURL = containerURL.appendingPathComponent("frame_metadata.json")
                if FileManager.default.fileExists(atPath: frameMetaURL.path) {
                    try FileManager.default.removeItem(at: frameMetaURL)
                }

                // Write initial IDLE state
                let initialMeta: [String: Any] = [
                    "status": "AWAITING_BROADCAST_START",
                    "prepared_at": ISO8601DateFormatter().string(from: Date())
                ]
                let data = try JSONSerialization.data(withJSONObject: initialMeta, options: .prettyPrinted)
                try data.write(to: sessionMetaURL, options: .atomic)

                promise.resolve(true)
            } catch {
                promise.reject("PREPARE_ERROR", error.localizedDescription)
            }
        }

        // MARK: - Cleanup

        /// Cleans up all frame data and metadata from the shared container.
        /// Call this after frames have been processed and uploaded.
        AsyncFunction("cleanupFrames") { (promise: Promise) in
            guard let containerURL = FileManager.default.containerURL(
                forSecurityApplicationGroupIdentifier: self.appGroupID
            ) else {
                promise.resolve(false)
                return
            }

            do {
                // Remove frames directory and recreate empty
                let framesDir = containerURL.appendingPathComponent("frames")
                if FileManager.default.fileExists(atPath: framesDir.path) {
                    try FileManager.default.removeItem(at: framesDir)
                    try FileManager.default.createDirectory(
                        at: framesDir,
                        withIntermediateDirectories: true
                    )
                }

                // Remove frame metadata
                let frameMetaURL = containerURL.appendingPathComponent("frame_metadata.json")
                if FileManager.default.fileExists(atPath: frameMetaURL.path) {
                    try FileManager.default.removeItem(at: frameMetaURL)
                }

                // Reset session metadata to IDLE
                let sessionMetaURL = containerURL.appendingPathComponent("session_metadata.json")
                let idleMeta: [String: Any] = [
                    "status": "IDLE",
                    "cleaned_at": ISO8601DateFormatter().string(from: Date())
                ]
                let data = try JSONSerialization.data(withJSONObject: idleMeta, options: .prettyPrinted)
                try data.write(to: sessionMetaURL, options: .atomic)

                promise.resolve(true)
            } catch {
                promise.reject("CLEANUP_ERROR", error.localizedDescription)
            }
        }

        /// Returns the total storage used by captured frames in bytes.
        Function("getStorageUsed") { () -> Int in
            guard let containerURL = FileManager.default.containerURL(
                forSecurityApplicationGroupIdentifier: self.appGroupID
            ) else {
                return 0
            }

            let framesDir = containerURL.appendingPathComponent("frames")
            guard let contents = try? FileManager.default.contentsOfDirectory(atPath: framesDir.path) else {
                return 0
            }

            var totalSize = 0
            for filename in contents where filename.hasSuffix(".jpg") {
                let filePath = framesDir.appendingPathComponent(filename).path
                if let attrs = try? FileManager.default.attributesOfItem(atPath: filePath),
                   let size = attrs[.size] as? Int {
                    totalSize += size
                }
            }
            return totalSize
        }
    }

    // MARK: - Private

    private func stopPolling() {
        statusPollTimer?.invalidate()
        statusPollTimer = nil
    }
}

File: AlgorithmLens_Cowork/mobile/modules/broadcast/android/BroadcastModule.kt
package com.algorithmlens.broadcast

import android.app.Activity
import android.content.Context
import android.content.Intent
import android.media.projection.MediaProjectionManager
import android.os.Build
import android.os.Handler
import android.os.Looper
import android.util.Base64
import expo.modules.kotlin.Promise
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.exception.CodedException
import org.json.JSONObject
import java.io.File
import java.util.Timer
import java.util.TimerTask

/**
 * BroadcastModule — Expo native module bridge for Android MediaProjection.
 *
 * This module provides the React Native interface to start/stop/query
 * Android MediaProjection screen capture. The capture runs in a foreground
 * service (MediaProjectionService) with a persistent notification.
 *
 * Responsibilities:
 * - Query MediaProjection availability
 * - Request screen capture permission via Activity result
 * - Start/stop the foreground capture service
 * - Read session metadata and frame data from AndroidSharedStorage
 * - Emit status change events to React Native
 * - Clean up storage after processing
 *
 * Key difference from iOS:
 * On iOS, the user taps RPSystemBroadcastPickerView to start recording.
 * On Android, we request MediaProjection permission via an Activity intent,
 * then start a foreground service. The permission flow is triggered
 * from the React Native layer via requestScreenCapture().
 */
class BroadcastModule : Module() {

    private val REQUEST_CODE_SCREEN_CAPTURE = 7001
    private var statusPollTimer: Timer? = null
    private val mainHandler = Handler(Looper.getMainLooper())

    private val sharedStorage: AndroidSharedStorage
        get() = AndroidSharedStorage(appContext.reactContext ?: throw CodedException("CONTEXT_ERROR", "React context not available", null))

    override fun definition() = ModuleDefinition {
        Name("ExpoBroadcast")

        // Emit events to JS when broadcast status changes
        Events("onStatusChange", "onFrameCountUpdate")

        // ── Availability & Status ─────────────────────────

        /**
         * Returns whether MediaProjection is available on this device.
         * Requires Android 5.0 (API 21)+.
         */
        Function("isAvailable") {
            return@Function Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP
        }

        /**
         * Returns the current broadcast session status from storage.
         */
        Function("getStatus") {
            val metadata = sharedStorage.readSessionMetadata()
            if (metadata != null) {
                return@Function jsonToMap(metadata)
            }
            return@Function mapOf("status" to "IDLE")
        }

        /**
         * Returns the storage directory path for debugging/diagnostics.
         */
        Function("getSharedContainerPath") {
            val context = appContext.reactContext ?: return@Function null
            return@Function File(context.filesDir, "broadcast").absolutePath
        }

        // ── Frame Data ────────────────────────────────────

        /**
         * Returns the number of unique JPEG frames stored.
         */
        Function("getFrameCount") {
            return@Function sharedStorage.frameCount()
        }

        /**
         * Returns the frame metadata array.
         */
        Function("getFrameMetadata") {
            val frames = sharedStorage.readFrameMetadata()
            val result = mutableListOf<Map<String, Any?>>()
            for (i in 0 until frames.length()) {
                result.add(jsonToMap(frames.getJSONObject(i)))
            }
            return@Function result
        }

        /**
         * Returns the full file paths for all captured JPEG frames, sorted.
         */
        Function("getFramePaths") {
            return@Function sharedStorage.allFramePaths()
        }

        /**
         * Reads a single frame's JPEG data as a base64-encoded string.
         */
        Function("getFrameBase64") { filename: String ->
            return@Function sharedStorage.getFrameBase64(filename)
        }

        // ── Session Management ────────────────────────────

        /**
         * Starts polling storage for status changes.
         * Emits onStatusChange and onFrameCountUpdate events.
         */
        Function("startStatusPolling") {
            stopPolling()

            var lastStatus = ""
            var lastFrameCount = 0

            statusPollTimer = Timer().also { timer ->
                timer.scheduleAtFixedRate(object : TimerTask() {
                    override fun run() {
                        val metadata = sharedStorage.readSessionMetadata() ?: return

                        val currentStatus = metadata.optString("status", "IDLE")
                        val currentFrames = metadata.optInt("frames_unique", 0)

                        if (currentStatus != lastStatus) {
                            lastStatus = currentStatus
                            mainHandler.post {
                                sendEvent("onStatusChange", mapOf(
                                    "status" to currentStatus,
                                    "metadata" to jsonToMap(metadata)
                                ))
                            }

                            // Auto-stop polling on terminal states
                            if (currentStatus == "COMPLETE" || currentStatus == "FAILED") {
                                stopPolling()
                            }
                        }

                        if (currentFrames != lastFrameCount) {
                            lastFrameCount = currentFrames
                            mainHandler.post {
                                sendEvent("onFrameCountUpdate", mapOf(
                                    "frameCount" to currentFrames,
                                    "framesTotal" to metadata.optInt("frames_captured", 0)
                                ))
                            }
                        }
                    }
                }, 0, 1000)
            }
        }

        /**
         * Stops polling for status changes.
         */
        Function("stopStatusPolling") {
            stopPolling()
        }

        /**
         * Prepares storage for a new broadcast session.
         */
        AsyncFunction("prepareSession") { promise: Promise ->
            try {
                sharedStorage.prepareForNewSession()
                promise.resolve(true)
            } catch (e: Exception) {
                promise.reject("PREPARE_ERROR", e.message ?: "Failed to prepare session", e)
            }
        }

        /**
         * Requests MediaProjection permission and starts the capture service.
         * This triggers the system permission dialog for screen capture.
         *
         * The permission result is handled via onActivityResult, which
         * starts the MediaProjectionService with the result token.
         */
        AsyncFunction("requestScreenCapture") { promise: Promise ->
            val activity = appContext.currentActivity
            if (activity == null) {
                promise.reject("NO_ACTIVITY", "No activity available to request screen capture", null)
                return@AsyncFunction
            }

            val projectionManager = activity.getSystemService(Context.MEDIA_PROJECTION_SERVICE) as? MediaProjectionManager
            if (projectionManager == null) {
                promise.reject("NO_PROJECTION", "MediaProjection not available on this device", null)
                return@AsyncFunction
            }

            // Store the promise to resolve when we get the activity result
            pendingCapturePromise = promise

            val captureIntent = projectionManager.createScreenCaptureIntent()
            activity.startActivityForResult(captureIntent, REQUEST_CODE_SCREEN_CAPTURE)
        }

        /**
         * Stops the screen capture service.
         */
        Function("stopCapture") {
            val context = appContext.reactContext ?: return@Function
            val stopIntent = Intent(context, MediaProjectionService::class.java).apply {
                action = MediaProjectionService.ACTION_STOP
            }
            context.startService(stopIntent)
        }

        // ── Cleanup ───────────────────────────────────────

        /**
         * Cleans up all frame data and metadata from storage.
         */
        AsyncFunction("cleanupFrames") { promise: Promise ->
            try {
                sharedStorage.fullReset()
                promise.resolve(true)
            } catch (e: Exception) {
                promise.reject("CLEANUP_ERROR", e.message ?: "Cleanup failed", e)
            }
        }

        /**
         * Returns the total storage used by captured frames in bytes.
         */
        Function("getStorageUsed") {
            return@Function sharedStorage.totalFrameSize()
        }

        // ── Activity Result Handling ──────────────────────

        OnActivityResult { _, payload ->
            val requestCode = payload.requestCode
            val resultCode = payload.resultCode
            val data = payload.data

            if (requestCode == REQUEST_CODE_SCREEN_CAPTURE) {
                if (resultCode == Activity.RESULT_OK && data != null) {
                    // User granted screen capture permission
                    val context = appContext.reactContext
                    if (context != null) {
                        val serviceIntent = Intent(context, MediaProjectionService::class.java).apply {
                            action = MediaProjectionService.ACTION_START
                            putExtra(MediaProjectionService.EXTRA_RESULT_CODE, resultCode)
                            putExtra(MediaProjectionService.EXTRA_RESULT_DATA, data)
                        }

                        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                            context.startForegroundService(serviceIntent)
                        } else {
                            context.startService(serviceIntent)
                        }
                    }
                    pendingCapturePromise?.resolve(true)
                } else {
                    // User denied screen capture
                    pendingCapturePromise?.reject(
                        "USER_DENIED",
                        "Screen capture permission was denied",
                        null
                    )
                }
                pendingCapturePromise = null
            }
        }
    }

    // ── Private ───────────────────────────────────────────

    private var pendingCapturePromise: Promise? = null

    private fun stopPolling() {
        statusPollTimer?.cancel()
        statusPollTimer = null
    }

    /**
     * Converts a JSONObject to a Map<String, Any?> for Expo module return values.
     */
    private fun jsonToMap(json: JSONObject): Map<String, Any?> {
        val map = mutableMapOf<String, Any?>()
        val keys = json.keys()
        while (keys.hasNext()) {
            val key = keys.next()
            val value = json.opt(key)
            map[key] = when (value) {
                JSONObject.NULL -> null
                is JSONObject -> jsonToMap(value)
                else -> value
            }
        }
        return map
    }
}
